crewai_tools/tools/contextual_tool/contextual_tool.py

```diff
@@ -0,0 +1,486 @@
+from typing import Any, Optional, Type, List, Union
+from crewai.tools import BaseTool
+from pydantic import BaseModel, Field
+import time
+import requests
+import warnings
+
+
+class ContextualQuerySchema(BaseModel):
+    """Schema for contextual query tool."""
+    query: str = Field(..., description="Query to send to the Contextual AI agent.")
+
+
+class ContextualTool(BaseTool):
+    """Tool to interact with Contextual AI RAG agents."""
+    
+    name: str = "Contextual AI Query Tool"
+    description: str = "Use this tool to query a Contextual AI RAG agent with access to your documents"
+    args_schema: Type[BaseModel] = ContextualQuerySchema
+    
+    api_key: str
+    agent_id: Optional[str] = None
+    datastore_id: Optional[str] = None
+    contextual_client: Any = None
+
+    def __init__(self, **kwargs):
+        super().__init__(**kwargs)
+        try:
+            from contextual import ContextualAI
+            self.contextual_client = ContextualAI(api_key=self.api_key)
+        except ImportError:
+            raise ImportError(
+                "contextual-client package is required. Install it with: pip install contextual-client"
+            )
+
+    def _check_documents_ready(self, datastore_id: str) -> bool:
+        """Check if all documents in datastore are ready (not pending)."""  
+        url = f"https://api.contextual.ai/v1/datastores/{datastore_id}/documents"
+        headers = {"Authorization": f"Bearer {self.api_key}"}
+        response = requests.get(url, headers=headers)
+        
+        if response.status_code == 200:
+            data = response.json()
+            documents = data.get('documents', [])
+            # print(f"Documents: {documents}")
+
+            # Return False if ANY document is still processing/pending
+            return not any(doc.get('status') in ('processing', 'pending') for doc in documents)
+        return True
+
+    def _run(self, query: str) -> str:
+        """Run the tool by querying the Contextual AI agent."""
+        if not self.agent_id:
+            raise ValueError("Agent ID is required to query the Contextual AI agent")
+        
+        if not self.datastore_id:
+            warnings.warn("Datastore ID not provided - skipping document readiness check", UserWarning)
+        else:
+            # Wait for documents to be ready
+            attempts = 0  
+            while not self._check_documents_ready(self.datastore_id) and attempts < 20:  
+                print(f"Waiting for documents to be processed... Attempt {attempts}/20, waiting for 30 seconds ...")  
+                time.sleep(30)  
```

Comment:
Avoid using `time.sleep`, as it can block execution and negatively impact parallel or asynchronous workflows. Consider using non-blocking alternatives such as asynchronous waits or event-based triggers to prevent delays in other processes running concurrently.